Energy resource

Phoscur opened the pull request:
Add new Entities: Factory, FactoryEnvironment
Add new ValueObjects: Building, Prosumer, Energy, EnergyCalculation

- Factory Energy ✔
- Add energy outage handling strategies ✔
- Tests, Coverage (keep it at 100%) ✔

Questionable:
Energy is still a copy from Resource, should it be the same class? Are the generics BuildingType and ResourceType useless or is it nice to use configurable TypeScript enums?

There are still many Todos left but the PR is already quite big and the next refactors are out of scope.
The next steps are to add BuildingProcesses and the main entity to encapsulate resources and buildings and exporting the engine package for a simple console UI.

Rebased and squashed after npm audit

miffels commented on src/Building.spec.ts:
```diff
@@ -1,27 +1,35 @@
 import { expect } from "chai";
 
+import { stock, requirements, prosumption } from "./examples";
```
Go, go, power rangers!

miffels commented on src/Building.ts:
```diff
@@ -1,22 +1,50 @@
+import {
```
Du kannst es tun, oh Sailor Moon!

miffels commented on src/BuildingRequirement.spec.ts:
```diff
@@ -4,30 +4,30 @@ import examples, { Types } from "./resources/examples";
 import BuildingRequirement from "./BuildingRequirement";
 import ResourceCollection from "./resources/ResourceCollection";
 import Stock from "./resources/Stock";
-import Empire from "./Empire";
```
Du wirst der allerbeste sein, wie keiner vor dir war!

miffels commented on src/BuildingRequirement.ts:
```diff
@@ -1,6 +1,5 @@
-import { ResourceIdentifier } from "./resources/Resource";
```
Sie lösen jeden Fall, was sie auch tun!

miffels commented on src/Factory.spec.ts:
```diff
@@ -0,0 +1,42 @@
+import { expect } from "chai";
```
Sie lösen jeden Fall, was sie auch tun,
das Böse hat nie Zeit sich auszuruhen.

miffels commented on src/Empire.ts:
```diff
@@ -1,18 +1,29 @@
-import { ResourceIdentifier } from "./resources/Resource";
```
Gummibär'n hüpfen hier und dort und überall!

Phoscur replied:
Überall anscheinend noch nicht, ist dir hier schon die Puste ausgegangen? Coverage ist wieder bei 100% :P

miffels reviewed (commented):
If that isn't motivational, I don't now what is.

Phoscur committed:
Add Energy as seperate ResourceType

Add readme
Add Consumer and EnergyCalculation
Add Factory
Add BuildingRequirement
Add Prosumer and example prosumption
Add ResourceLike
Add FactoryEnvironment

Properly test prosumer recalculation strategy failure
Add rebalancing for resource production on low energy & over consumption: halt buildings out of stock

Add upgrade- & downgrade- time & -cost

Phoscur committed:
Add resources.productionTable

Refactor rename previous method to "entries" for toString usage

Phoscur committed:
Fix resource recalculation strategy

Phoscur committed:
Add documentation
